File: paddock/data/fixtures/okf-behavior-ts/oracle/service.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";
import { dispatch } from "./service.ts";

type Case = {
  name: string;
  items: string[];
  sent: string[];
  limit: number;
  status: string;
  count: number;
  after: string[];
  error: string;
};

const cases: Case[] = [
  { name: "success", items: ["a", "b", "c"], sent: ["z"], limit: 2, status: "sent", count: 2, after: ["z", "a", "b"], error: "" },
  { name: "empty", items: [], sent: ["z"], limit: 2, status: "empty", count: 0, after: ["z"], error: "" },
  { name: "error", items: ["a"], sent: ["z"], limit: -1, status: "", count: 0, after: ["z"], error: "limit must be nonnegative" },
  { name: "empty-error", items: [], sent: ["z"], limit: -1, status: "", count: 0, after: ["z"], error: "limit must be nonnegative" },
  { name: "zero", items: ["a", "b"], sent: [], limit: 0, status: "sent", count: 0, after: [], error: "" },
  { name: "bounded", items: ["a", "b"], sent: [], limit: 5, status: "sent", count: 2, after: ["a", "b"], error: "" },
];

for (const tc of cases) {
  test(`dispatch contract/${tc.name}`, () => {
    const sent = [...tc.sent];
    if (tc.error) {
      assert.throws(() => dispatch(tc.items, sent, tc.limit), { name: "Error", message: tc.error });
    } else {
      const result = dispatch(tc.items, sent, tc.limit);
      assert.equal(result.status, tc.status);
      assert.equal(result.count, tc.count);
    }
    assert.deepEqual(sent, tc.after);
  });
}

File: paddock/data/fixtures/okf-behavior-go/oracle/service_test.go
package dispatch

import (
	"reflect"
	"testing"
)

func TestDispatchContract(t *testing.T) {
	for _, tc := range []struct {
		name string
		items []string
		limit int
		status string
		count int
		want []string
		err string
	}{
		{"success", []string{"a", "b", "c"}, 2, "sent", 2, []string{"existing", "a", "b"}, ""},
		{"empty", nil, 2, "empty", 0, []string{"existing"}, ""},
		{"error", []string{"a"}, -1, "", 0, []string{"existing"}, "limit must be nonnegative"},
		{"empty-error", nil, -1, "", 0, []string{"existing"}, "limit must be nonnegative"},
		{"zero", []string{"a"}, 0, "sent", 0, []string{"existing"}, ""},
		{"bounded", []string{"a"}, 9, "sent", 1, []string{"existing", "a"}, ""},
	} {
		t.Run(tc.name, func(t *testing.T) {
			sent := []string{"existing"}
			status, count, err := Dispatch(tc.items, &sent, tc.limit)
			if status != tc.status || count != tc.count || !reflect.DeepEqual(sent, tc.want) {
				t.Fatalf("got status=%q count=%d sent=%v; want %q %d %v", status, count, sent, tc.status, tc.count, tc.want)
			}
			if tc.err == "" && err != nil || tc.err != "" && (err == nil || err.Error() != tc.err) {
				t.Fatalf("got error %v; want %q", err, tc.err)
			}
		})
	}
}
